libs/custom-paper-handler/src/driver/scanner_capability.ts

```diff
@@ -0,0 +1,191 @@
+  0x01: 'BW',
+  0x02: 'grayscale',
+  0x03: 'RGB',
+  0x04: 'RGBU',
+};
+const ResolutionDecoder: Decoder<ResolutionCapabilityOption> = {
+  0x01: 50,
+  0x02: 100,
+  0x03: 150,
+  0x04: 200,
+  0x05: 250,
+  0x06: 300,
+  0x07: 600,
+};
+
+function decode<Capability>(
+  code: number,
+  decoder: Decoder<Capability>
+): Capability {
+  const option = decoder[code];
+  if (!option) {
+    throw new Error(`Unrecognized capability option code ${code} for `);
+  }
+  return option;
+}
+
+function getEmptyScannerCapability(): ScannerCapability {
+  return {
+    paperMovementAfterScanOptions: [],
+    paperSensors: [],
+    scannerSensors: [],
+    scanSources: [],
+    scanLightOptions: [],
+    scanDataFormats: [],
+    horizontalResolutions: [],
+    verticalResolutions: [],
+    maxDotsHorizontal: 0,
+    maxDotsVertical: 0,
+    imageBufferSize: 0,
+    transactionBufferSize: 0,
+  };
+}
+
+export function parseScannerCapability(data: DataView): ScannerCapability {
+  const scannerCapability = getEmptyScannerCapability();
+  let byteIndex = 4;
+
+  // Parses capability block starting at the current byte
+  function parseOptions<T>(decoder: Decoder<T>): T[] {
+    const options: T[] = [];
+    const numOptions = data.getUint8(byteIndex + 1);
+    for (let i = 0; i < numOptions; i += 1) {
+      options.push(decode(data.getUint8(byteIndex + 2 + i), decoder));
+    }
+    return options;
+  }
+
+  while (byteIndex < data.byteLength) {
+    const capabilityId = data.getUint8(byteIndex);
+    switch (capabilityId) {
+      case 0x80:
+        scannerCapability.paperMovementAfterScanOptions = parseOptions(
+          PaperMovementAfterScanDecoder
+        );
+        break;
+      case 0x81:
+        scannerCapability.paperSensors = parseOptions(PaperSensorDecoder);
+        break;
+      case 0x82:
+        scannerCapability.scannerSensors = parseOptions(ScannerSensorDecoder);
+        break;
+      case 0x83:
+        scannerCapability.scanSources = parseOptions(ScanSourcesDecoder);
+        break;
+      case 0x84:
+        scannerCapability.scanLightOptions = parseOptions(ScanLightsDecoder);
+        break;
+      case 0x85:
+        scannerCapability.scanDataFormats = parseOptions(ScanDataFormatDecoder);
+        break;
+      case 0x86:
+        scannerCapability.horizontalResolutions =
+          parseOptions(ResolutionDecoder);
+        break;
+      case 0x87:
+        scannerCapability.verticalResolutions = parseOptions(ResolutionDecoder);
+        break;
+      case 0x88:
+        scannerCapability.maxDotsHorizontal = data.getUint32(byteIndex + 3);
+        scannerCapability.maxDotsVertical = data.getUint32(byteIndex + 8);
+        break;
+      case 0x89:
+        scannerCapability.imageBufferSize = data.getUint32(byteIndex + 3);
+        scannerCapability.transactionBufferSize = data.getUint32(byteIndex + 8);
+        break;
+      default:
+        throw new Error('unknown capability ID');
+    }
```

Comment:
This can hopefully be replaced by [`oneOf`](https://github.com/votingworks/vxsuite/pull/3481).

Reply:
Thank you for opening this!